### PR TITLE
Bootloader ISO EFI Support

### PR DESCRIPTION
**Pull request by klaases**
Fixes: #13122

This PR is an important prerequisite for building minikube ISO on ARM64.

Note: the `containerd` version has been lowered 1.4.9, as 1.5.8 has a rotating hash issue that prevents the ISO from building properly. For more information, see: https://github.com/kubernetes/minikube/issues/13193 

Note: make sure to merge #12892 before submission. [CONFIRMED]

---

Build the ISO, see https://minikube.sigs.k8s.io/docs/contrib/building/iso/

Use `mk ssh` to ssh into minikube ISO.

Run `[ -d /sys/firmware/efi ] && echo UEFI || echo BIOS`

**Before:**

`BIOS`

**After:**

`UEFI`

---

Have tested the new ISO with the following commands.

Download the latest ISO, and place in `minikube/out` folder.
`curl -LO 
https://storage.googleapis.com/minikube-builds/iso/13123/minikube-v1.24.0-1640194376-13123.iso`

Alternative URL:
`gs://minikube-builds/iso/13123/minikube-v1.24.0-1640194376-13123.iso`

Run `make` from minikube home folder.

Run `./out/minikube start --iso-url=file://$(pwd)/out/minikube-v1.24.0-1640194376-13123.iso` to start minikube with the new ISO.

Use `mk ssh` to ssh into minikube ISO.

Run `[ -d /sys/firmware/efi ] && echo UEFI || echo BIOS` to confirm that minikube is running with the UEFI bootloader.

### CODE REVIEW

**Comment by klaases** [comment]
wip:add efi bootloader

**Comment by klaases** [comment]
This configuration resulted in the following error when building ISO:

```
g++: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-7/README.Bugs> for instructions.
Makefile:1116: recipe for target 'gimple-match.o' failed
make[4]: *** [gimple-match.o] Error 4
make[4]: *** Waiting for unfinished jobs....
rm gcc.pod
make[4]: Leaving directory '/mnt/out/buildroot/output/build/host-gcc-initial-9.4.0/build/gcc'
Makefile:4300: recipe for target 'all-gcc' failed
make[3]: *** [all-gcc] Error 2
make[3]: Leaving directory '/mnt/out/buildroot/output/build/host-gcc-initial-9.4.0/build'
package/pkg-generic.mk:247: recipe for target '/mnt/out/buildroot/output/build/host-gcc-initial-9.4.0/.stamp_built' failed
make[2]: *** [/mnt/out/buildroot/output/build/host-gcc-initial-9.4.0/.stamp_built] Error 2
make[2]: Leaving directory '/mnt/out/buildroot'
make[1]: *** [minikube_iso] Error 2
Makefile:286: recipe for target 'minikube_iso' failed
make[1]: Leaving directory '/mnt'
make: *** [out/minikube.iso] Error 2
Makefile:312: recipe for target 'out/minikube.iso' failed
make: *** [out/minikube.iso] Error 2
```

**Comment by klaases** [comment]
Added additional files for testing.

**Comment by medyagh** [comment]
lets ensure this ISO change that was merged gets added to the PR

https://github.com/kubernetes/minikube/pull/12892

File: deploy/iso/minikube-iso/configs/minikube_defconfig [hunk 1]
```diff
@@ -1,10 +1,53 @@
...
+# Bootloader
+BR2_TARGET_GRUB2=y
+BR2_TARGET_GRUB2_X86_64_EFI=y
+BR2_TARGET_GRUB2_BUILTIN_CONFIG="$(BR2_EXTERNAL_MINIKUBE_PATH)/board/iso/x86_64/grub-stub.cfg"
+BR2_TARGET_GRUB2_BUILTIN_MODULES="boot linux ext2 fat squash4 part_msdos part_gpt normal efi_gop udf iso9660 search search_fs_file search_fs_uuid search_label"
+
+# Filesystem / image
+BR2_TARGET_ROOTFS_CPIO=y
+BR2_TARGET_ROOTFS_CPIO_GZIP=y
+# BR2_TARGET_ROOTFS_TAR is not set
```
Comment (spowelljr) [inline deploy/iso/minikube-iso/configs/minikube_defconfig:24]: Can we just delete this or is this a note for later?

Comment (klaases) [reply deploy/iso/minikube-iso/configs/minikube_defconfig]: This was part of the original default configuration for x86_64 with EFI, but I am fine to remove it.

Also removed a few other commented lines.

**Comment by sharifelgamal** [comment]
ok-to-build-iso

**Comment by klaases** [comment]
Error based on building new ISO includes:

```
ERROR: v1.5.8.tar.gz has wrong sha256 hash:
ERROR: expected: a41ab8d39393c9456941b477c33bb1b221a29b635f1c9a99523aab2f5e74f790
ERROR: got     : 0890f7b0ee8e20a279a617c60686874b3c7a99e064adb2b38d884499b5284c43
ERROR: Incomplete download, or man-in-the-middle (MITM) attack
--2021-12-16 19:31:59--  http://sources.buildroot.net/containerd-bin/v1.5.8.tar.gz
Resolving sources.buildroot.net (sources.buildroot.net)... 172.67.72.56, 104.26.1.37, 104.26.0.37, ...
Connecting to sources.buildroot.net (sources.buildroot.net)|172.67.72.56|:80... connected.
HTTP request sent, awaiting response... 404 Not Found
2021-12-16 19:31:59 ERROR 404: Not Found.

--2021-12-16 19:31:59--  http://sources.buildroot.net/v1.5.8.tar.gz
Resolving sources.buildroot.net (sources.buildroot.net)... 172.67.72.56, 104.26.1.37, 104.26.0.37, ...
Connecting to sources.buildroot.net (sources.buildroot.net)|172.67.72.56|:80... connected.
HTTP request sent, awaiting response... 404 Not Found
2021-12-16 19:31:59 ERROR 404: Not Found.

make[1]: *** [package/pkg-generic.mk:148: /home/jenkins/go/src/k8s.io/minikube/out/buildroot/output/build/containerd-bin-v1.5.8/.stamp_downloaded] Error 1
make[1]: Leaving directory '/home/jenkins/go/src/k8s.io/minikube/out/buildroot'
make: *** [Makefile:290: minikube_iso] Error 2
```

**Comment by klaases** [comment]
I am attributing the ISO build failure to an incomplete download of containerd, we will try it again :)

**Comment by klaases** [comment]
ok-to-build-iso

**Comment by klaases** [comment]
Based on the error that occurred during the most recent run:

```
/home/jenkins/go/src/k8s.io/minikube/deploy/iso/minikube-iso/board/iso/x86_64/post-image.sh: 31: /home/jenkins/go/src/k8s.io/minikube/deploy/iso/minikube-iso/board/iso/x86_64/post-image.sh: mkisofs: not found
make[1]: *** [Makefile:836: target-post-image] Error 127
```

```
Installed `mkisofs` to Debian VM in GCP

- Run “sudo apt-get update”
- Run “sudo apt-get install genisoimage”
- Run “which genisoimage” to confirm.
- Run “sudo ln -s /usr/bin/genisoimage /usr/bin/mkisofs” to establish symlink.
- Run “ls -lah $(which mkisofs)” confirm symlink.

```

**Comment by klaases** [comment]
ok-to-build-iso

**Comment by afbjorklund** [comment]
`genisoimage: unrecognized option '-eltorito-platform=efi'`

**Comment by afbjorklund** [comment]
> I am attributing the ISO build failure to an incomplete download of containerd, we will try it again :)

@klaases containerd checksum issues added as:
* #13193 

https://github.com/kubernetes/minikube/pull/13123#issuecomment-996169421

**Comment by klaases** [comment]
> @klaases lets ensure this ISO change that was merged gets added to the PR
> 
> #12892
> 
> only one line needed to be pulled from that PR and that is:
> 
> ```
> CONFIG_IP_SCTP=y
> ```

Yes, this change will be included in this PR.

**Comment by klaases** [comment]
> https://storage.cloud.google.com/minikube-builds/logs/13123/da7f85a/iso_build.txt

Glad to have this as confirmation. Had updated the sha256 but we see it continue to change during each build.

**Comment by afbjorklund** [comment]
containerd also breaks everywhere else, that is checking the checksum. only difference is if they use sha256 or sha512

**Comment by klaases** [comment]
Try to build UEFI ISO with downgraded version of containerd, due to hashing issue:
https://github.com/kubernetes/minikube/issues/13193

**Comment by klaases** [comment]
ok-to-build-iso

**Comment by sharifelgamal** [comment]
ok-to-build-iso

**Comment by klaases** [comment]
Trying to build on a new ubuntu server, rather than trying with Docker.

Trying again with `mkisofs` installed on new server.

```
18:41:49 /home/jenkins/go/src/k8s.io/minikube/deploy/iso/minikube-iso/board/iso/x86_64/post-image.sh: 31: /home/jenkins/go/src/k8s.io/minikube/deploy/iso/minikube-iso/board/iso/x86_64/post-image.sh: mkisofs: not found
```

**Comment by klaases** [comment]
ok-to-build-iso

**Comment by sharifelgamal** [comment]
ok-to-build-iso

**Comment by klaases** [comment]
Nest steps will be to install `gsutil`:

https://cloud.google.com/storage/docs/gsutil_install#deb

```
make: gsutil: Command not found
make: *** [release-iso] Error 127
```

**Comment by sharifelgamal** [comment]
ok-to-build-iso

**Comment by sharifelgamal** [comment]
ok-to-build-iso

**Comment by klaases** [comment]
Have tested the new ISO with the following commands.

Download the latest ISO, and place in `minikube/out` folder.
`curl -LO 
https://storage.googleapis.com/minikube-builds/iso/13123/minikube-v1.24.0-1640194376-13123.iso`

Alternative URL:
`gs://minikube-builds/iso/13123/minikube-v1.24.0-1640194376-13123.iso`

Run `make` from minikube home folder.

Run `./out/minikube start --iso-url=file://$(pwd)/out/minikube-v1.24.0-1640194376-13123.iso` to start minikube with the new ISO.

Use `mk ssh` to ssh into minikube ISO.

Run `[ -d /sys/firmware/efi ] && echo UEFI || echo BIOS` to confirm that minikube is running with the UEFI bootloader.

**Comment by afbjorklund** [comment]
@klaases we can upgrade containerd to 1.4.12, with https://github.com/kubernetes/minikube/pull/13195

that would also address the security issue (CVE-2021-41190)

**Review by sharifelgamal — APPROVED** [review]

**Pull request merged**